homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -101,76 +202,204 @@ async def async_step_zeroconf(self, discovery_info):
-                        unique_id = self.mac
-                        await self.async_set_unique_id(
-                            unique_id, raise_on_progress=False
-                        )
-                        self._abort_if_unique_id_configured()
-                        return self.async_create_entry(
-                            title=DEFAULT_GATEWAY_NAME,
-                            data={
-                                CONF_FLOW_TYPE: CONF_GATEWAY,
-                                CONF_HOST: self.host,
-                                CONF_TOKEN: token,
-                                CONF_MODEL: model,
-                                CONF_MAC: self.mac,
-                            },
-                        )
-
-                # Setup all other Miio Devices
-                name = user_input.get(CONF_NAME, model)
-
-                for device_model in MODELS_ALL_DEVICES:
-                    if model.startswith(device_model):
-                        unique_id = self.mac
-                        await self.async_set_unique_id(
-                            unique_id, raise_on_progress=False
-                        )
-                        self._abort_if_unique_id_configured()
-                        return self.async_create_entry(
-                            title=name,
-                            data={
-                                CONF_FLOW_TYPE: CONF_DEVICE,
-                                CONF_HOST: self.host,
-                                CONF_TOKEN: token,
-                                CONF_MODEL: model,
-                                CONF_MAC: self.mac,
-                            },
-                        )
-                errors["base"] = "unknown_device"
-            else:
-                errors["base"] = "cannot_connect"
+            return await self.async_step_connect()
 
         if self.host:
             schema = vol.Schema(DEVICE_SETTINGS)
         else:
             schema = DEVICE_CONFIG
 
-        if errors:
-            schema = schema.extend(DEVICE_MODEL_CONFIG)
+        return self.async_show_form(step_id="manual", data_schema=schema, errors=errors)
+
+    async def async_step_connect(self, user_input=None):
+        """Connect to a xiaomi miio device."""
+        errors = {}
+        if self.host is None or self.token is None:
+            return self.async_abort(reason="incomplete_info")
+
+        # Try to connect to a Xiaomi Device.
+        connect_device_class = ConnectXiaomiDevice(self.hass)
+        await connect_device_class.async_connect_device(self.host, self.token)
+        device_info = connect_device_class.device_info
+
+        if self.model is None and device_info is not None:
+            self.model = device_info.model
 
-        return self.async_show_form(step_id="device", data_schema=schema, errors=errors)
+        if self.model is None:
+            errors["base"] = "cannot_connect"
+            return await self.async_step_model(errors=errors)
```

Comment:
Don't pass errors from outside a step method. The errors are specific to the form of the step.